File: app/src/main/java/soup/movie/ui/detail/DetailViewState.java
package soup.movie.ui.detail;

import java.util.List;

import soup.movie.data.model.TimeTable;
import soup.movie.data.model.Trailer;
import soup.movie.ui.BaseViewState;

interface DetailViewState extends BaseViewState {

    class LoadingState implements DetailViewState {

        private boolean mTheaterNotExists;

        LoadingState() {
            this(true);
        }

        LoadingState(boolean theaterExists) {
            mTheaterNotExists = theaterExists;
        }

        public boolean isTheaterNotExists() {
            return mTheaterNotExists;
        }

        @Override
        public String toString() {
            return "LoadingState{}";
        }
    }

    class DoneState implements DetailViewState {

        private TimeTable timeTable;
        private List<Trailer> trailers;

        DoneState(TimeTable timeTable, List<Trailer> trailers) {
            this.timeTable = timeTable;
            this.trailers = trailers;
        }

        public TimeTable getTimeTable() {
            return timeTable;
        }

        public List<Trailer> getTrailers() {
            return trailers;
        }

        @Override
        public String toString() {
            return "DoneState{" +
                    "timeTable=" + timeTable +
                    ", trailers=" + trailers +
                    '}';
        }
    }
}

File: app/src/main/java/soup/movie/ui/detail/DetailViewState.kt
package soup.movie.ui.detail

import android.net.Uri
import androidx.annotation.Keep
import androidx.annotation.LayoutRes
import soup.movie.data.model.Movie
import soup.movie.data.model.TheaterType
import soup.movie.data.model.Trailer

sealed class DetailViewState {

    @Keep
    object LoadingState : DetailViewState() {

        override fun toString(): String = javaClass.simpleName
    }

    @Keep
    data class DoneState(
            val items: List<ListItem>) : DetailViewState()

    @Keep
    data class ListItem(
            val type: TheaterType,
            @LayoutRes
            val layoutRes: Int,
            val movie: Movie,
            val trailers: List<Trailer> = emptyList())
}

class ShareAction(
    val imageUri: Uri,
    val mimeType: String
)

